### PR TITLE
Add rkey_ptr to self transport, tag matching performance improvement

### PR DESCRIPTION
**Pull request by roiedanino**
## What
Adding support for put_zcopy and get_zcopy on self transport (uct)

## Why ?
To avoid fallbacks to other transports when possible (proto v2).

## How ?
Implemented the API interface functions for put_zcopy and get_zcopy using memcpy, treating remote address as a local pointer

### CODE REVIEW

File: src/uct/sm/self/self.c [hunk 1]
```diff
@@ -357,6 +359,27 @@ ssize_t uct_self_ep_am_bcopy(uct_ep_h tl_ep, uint8_t id,
     return length;
 }
 
+ucs_status_t uct_self_ep_put_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov,
+                                   size_t iovcnt, uint64_t remote_addr,
+                                   uct_rkey_t rkey, uct_completion_t *comp)
+{
+    uct_self_ep_t UCS_V_UNUSED *ep = ucs_derived_of(tl_ep, uct_self_ep_t);
+    memcpy((void*)(rkey + remote_addr), iov->buffer, iov->length);
```
Comment (yosefe) [inline src/uct/sm/self/self.c:367]: 1. need to copy all IOVs, not just the first one
2. space line before

File: src/uct/sm/self/self.c [hunk 2]
```diff
@@ -357,6 +359,27 @@ ssize_t uct_self_ep_am_bcopy(uct_ep_h tl_ep, uint8_t id,
...
+    UCT_TL_EP_STAT_OP(&ep->super, PUT, ZCOPY, iov->length);
+    return UCS_OK;
+}
+
+ucs_status_t uct_self_ep_get_zcopy(uct_ep_h tl_ep, const uct_iov_t *iov,
+                                   size_t iovcnt, uint64_t remote_addr,
+                                   uct_rkey_t rkey, uct_completion_t *comp)
+{
+    uct_self_ep_t UCS_V_UNUSED *ep = ucs_derived_of(tl_ep, uct_self_ep_t);
+    memcpy(iov->buffer, (void*)(rkey + remote_addr), iov->length);
```
Comment (yosefe) [inline src/uct/sm/self/self.c:377]: 1. need to copy all IOVs, not just the first one
2. space line before

File: src/uct/sm/self/self.c [hunk 3]
```diff
@@ -357,6 +359,27 @@ ssize_t uct_self_ep_am_bcopy(uct_ep_h tl_ep, uint8_t id,
...
+    UCT_TL_EP_STAT_OP(&ep->super, GET, ZCOPY, iov->length);
+    return UCS_OK;
+}
+
```
Comment (yosefe) [inline src/uct/sm/self/self.c:381]: remove one space line

File: src/uct/sm/base/sm_md.c [hunk 4]
```diff
@@ -0,0 +1,9 @@
+#include "sm_md.h"
+
+ucs_status_t uct_sm_rkey_ptr(uct_component_t *component, uct_rkey_t rkey,
+                             void *handle, uint64_t raddr, void **laddr_p)
+{
+    /* rkey stores offset from the remote va */
+    *laddr_p = UCS_PTR_BYTE_OFFSET(raddr, (ptrdiff_t)rkey);
+    return UCS_OK;
+}
```
Comment (yosefe) [inline src/uct/sm/base/sm_md.c:9]: add newline

File: src/uct/sm/base/sm_md.c [hunk 5]
```diff
@@ -0,0 +1,9 @@
+#include "sm_md.h"
```
Comment (yosefe) [inline src/uct/sm/base/sm_md.c:1]: add copyright before, space line after

File: src/uct/sm/base/sm_md.h [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+#define UCT_SM_MD_H_
+
+#include <uct/base/uct_md.h>
+#include <ucs/config/types.h>
+#include <ucs/type/status.h>
+
+ucs_status_t uct_sm_rkey_ptr(uct_component_t *component, uct_rkey_t rkey,
+                             void *handle, uint64_t raddr, void **laddr_p);
+
+#endif
```
Comment (yosefe) [inline src/uct/sm/base/sm_md.h:19]: add newline

File: src/uct/sm/self/self.c [hunk 7]
```diff
@@ -121,7 +125,7 @@ static ucs_status_t uct_self_iface_query(uct_iface_h tl_iface, uct_iface_attr_t
     attr->cap.am.max_iov          = SIZE_MAX;
 
     attr->latency                 = UCS_LINEAR_FUNC_ZERO;
-    attr->bandwidth.dedicated     = 6911.0 * UCS_MBYTE;
+    attr->bandwidth.dedicated     = 13000 * UCS_MBYTE;//12180.0 * UCS_MBYTE;
```
Comment (yosefe) [inline src/uct/sm/self/self.c:128]: revert

File: src/uct/sm/self/self.c [hunk 8]
```diff
@@ -81,7 +82,10 @@ static ucs_status_t uct_self_iface_query(uct_iface_h tl_iface, uct_iface_attr_t
                                    UCT_IFACE_FLAG_ATOMIC_CPU       |
                                    UCT_IFACE_FLAG_PENDING          |
                                    UCT_IFACE_FLAG_CB_SYNC          |
-                                   UCT_IFACE_FLAG_EP_CHECK;
+                                   UCT_IFACE_FLAG_EP_CHECK         |
+                                   UCT_IFACE_FLAG_TAG_RNDV_ZCOPY   |
+                                   UCT_IFACE_FLAG_PUT_ZCOPY        |
+                                   UCT_IFACE_FLAG_GET_ZCOPY;
```
Comment (yosefe) [inline src/uct/sm/self/self.c:88]: pls remove pu_zcopy,get_zcopy
keep only rkey_ptr

File: src/uct/sm/base/sm_md.c [hunk 9]
```diff
@@ -0,0 +1,19 @@
+/**
+* Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2001-2019. ALL RIGHTS RESERVED.
```
Comment (yosefe) [inline src/uct/sm/base/sm_md.c:2]: 2023

File: src/uct/sm/base/sm_md.h [hunk 10]
```diff
@@ -0,0 +1,16 @@
+/**
+* Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2001-2019. ALL RIGHTS RESERVED.
```
Comment (yosefe) [inline src/uct/sm/base/sm_md.h:2]: 2023

File: src/ucp/rma/get_offload.c [hunk 11]
```diff
@@ -137,6 +137,7 @@ ucp_proto_get_offload_zcopy_send_func(ucp_request_t *req,
                                lpriv->super.md_index, UCP_DT_MASK_CONTIG_IOV,
                                next_iter, &iov, 1);
 
+    /* coverity[tainted_data_downcast] */
```
Comment (yosefe) [inline src/ucp/rma/get_offload.c:140]: is it still needed?

File: src/uct/sm/base/sm_md.c [hunk 12]
```diff
@@ -0,0 +1,19 @@
+/**
+* Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2001-2023. ALL RIGHTS RESERVED.
+* See file LICENSE for terms.
+*/
```
Comment (yosefe) [inline src/uct/sm/base/sm_md.c:4]: copy/paste from https://github.com/openucx/ucx/blob/4f8240871d11bc53bbb609263d32d0127be0dc57/test/apps/test_no_cuda_ctx.c#L2

File: src/uct/sm/self/self.c [hunk 13]
```diff
@@ -81,7 +82,8 @@ static ucs_status_t uct_self_iface_query(uct_iface_h tl_iface, uct_iface_attr_t
                                    UCT_IFACE_FLAG_ATOMIC_CPU       |
                                    UCT_IFACE_FLAG_PENDING          |
                                    UCT_IFACE_FLAG_CB_SYNC          |
-                                   UCT_IFACE_FLAG_EP_CHECK;
+                                   UCT_IFACE_FLAG_EP_CHECK         |
+                                   UCT_IFACE_FLAG_TAG_RNDV_ZCOPY;
```
Comment (yosefe) [inline src/uct/sm/self/self.c:86]: remove UCT_IFACE_FLAG_TAG_RNDV_ZCOPY

File: src/uct/sm/self/self.c [hunk 14]
```diff
@@ -462,7 +464,7 @@ static uct_component_t uct_self_component = {
     },
     .cm_config          = UCS_CONFIG_EMPTY_GLOBAL_LIST_ENTRY,
     .tl_list            = UCT_COMPONENT_TL_LIST_INITIALIZER(&uct_self_component),
-    .flags              = 0,
+    .flags              = UCT_MD_FLAG_RKEY_PTR | UCT_MD_FLAG_NEED_RKEY,
```
Comment (yosefe) [inline src/uct/sm/self/self.c:467]: remove UCT_MD_FLAG_NEED_RKEY

File: test/gtest/ucp/test_ucp_wireup.cc [hunk 15]
```diff
@@ -737,6 +737,9 @@ UCS_TEST_P(test_ucp_wireup_1sided, send_disconnect_reply2) {
 }
 
 UCS_TEST_P(test_ucp_wireup_1sided, disconnect_nb_onesided) {
+    // On loopback sender and receiver have the same ep,
+    // therefore it is impossible to disconnect only one of them.
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_wireup.cc:741]: is it still needed?

Comment (roiedanino) [reply test/gtest/ucp/test_ucp_wireup.cc:741]: Not anymore

**Review by yosefe — DISMISSED** [review]

**Comment by yosefe** [comment]
@gleon99 can you pls take a look as well?

**Review by gleon99 — DISMISSED** [review]

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@roiedanino pls update the commit message and PR title according to the code changes

**Review by yosefe — APPROVED** [review]

**Pull request merged**